### PR TITLE
PRVT-6: implement auth storage for session and token values

### PR DESCRIPTION
**Pull request by virdesai**
implement auth storage for session and token values with accessing values using local cache as priority vs the constructor based storage/localStorage

### CODE REVIEW

**Comment by loscm** [comment]
Requested review from a random `@lifeomic/security` team member, @cfbevan, due to term: **auth**.

This review is not blocking and is for broader awareness.  Consider if this change requires deeper security review and ask when it is necessary.

- [Automation](https://github.com/lifeomic/pr-security-flag-scanner/blob/master/src/lambdas/pr-security-flag-scanner/index.ts) | [#pr-flagged-sec-review](https://lifeomic.slack.com/messages/pr-flagged-sec-review)

File: src/utils.ts
Comment (aecorredor) [inline src/utils.ts:1]: could we place this inside a `utils` folder scoped to a more specific file?

**Review by aecorredor — COMMENTED** [review]
I know some of this was already in place before, but looking at the `APIBasedAuth` again, my sentiment was that we could maybe type it in a way that allows us to not have so much casting. If we have the context of `Token` and `Session` storage, that also means that we could probably have other types of storage later on. Couldn't we make this more generic (no specific casting based on a string) and separate the storage getters/setters from the actual shape of the objects we're dealing with?

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:102]: nbd we can, there are 2 promises within the session flow and 3 within token so this double await means both of the session removals execute and then all 3 of the token ones do. with the promise all in parallel we would have session_1 -> token_1 -> session_2 -> token_2 -> token_3

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:25]: safe? not entirely cause the potential exists where this is being used in a react-native context and no storage is provide and in that case, window.localStorage isn't a thing. so I could not use storage if nothing is provided. otherwise we'd have to add some sort of platform selector in which adds the react-native lib or requires a type on initialization to let us know what platform we're dealing with and if we can use window.localStorage as a default fallback

Comment (smolster) [reply src/APIBasedAuth.ts:25]: The existing `LOAuth` client already:
- defines a highly generalized type for a "Storage" abstraction,
- accepts it in the constructor, and
- has a pattern for using it internally.

Any reason not to just re-use the types + patterns defined there?

I think it's better to keep any platform-specific code out of the class, and just require providing a `storage` parameter.

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:232]: such that we can use 1 function and infer the type. but I'll look into making the entire storage concept generic without casting based on your comments. there will need to be some things that are specific tho to sessions and tokens and giving the user of app-tools the ability to set what the key name is that for the key-value store is important

File: src/APIBasedAuth.ts
Comment (jkdowdle) [reply src/APIBasedAuth.ts:25]: If it helps, the reason that each token had its own key is that on mobile we were hitting a size limit warning that might be treated as throwing an error in a future version of expos secure store https://docs.expo.dev/versions/latest/sdk/securestore/

Comment (virdesai) [reply src/APIBasedAuth.ts:25]: @smolster @aecorredor so theoretically we need to keep that known issue for expo in mind for the solution we make. so having a generic single function for setting storage which won't allow us to json stringify the tokens object and store it. knowing this, thoughts on the solution I have with individual key-value setting for everything?

File: src/APIBasedAuth.ts
Comment (aecorredor) [inline src/APIBasedAuth.ts:165]: Couldn't we make this `T` extend `session | token` to we can avoid the `as` casting below? The same would apply for other parts in this class where we do this. I think that should work from the way that `StorageType` is typed. If we can do that, maybe the class as is is type-safe enough and wouldn't require casting at all.

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:126]: I don't see any call-sites where we do `_getFromStorage('token')` -- am I missing something?

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:178]: This default (`${type}.${key}`) seems pretty open, and relatively likely to collide with other keys. Can we add some sort of LifeOmic-specific prefix? In `LOAuth`, we used `lo-app-tools-auth`, for example.

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:126]: you're correct but the functionality exists for the potential this class could be expanded to potentially need it

Comment (virdesai) [reply src/APIBasedAuth.ts:126]: woah, why 100? it's already covering 100% for functions. 95% for statements/lines. ~88% for branches

Comment (virdesai) [reply src/APIBasedAuth.ts]: @smolster alright removed the unused stuff for now and we can add to this in the future if/when needed

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:146]: @aecorredor removed some of the casting

File: src/APIBasedAuth.ts
Comment (smolster) [reply src/APIBasedAuth.ts:28]: @virdesai I forgot about the scenario of wanting to store the `session` + token in different places.

File: jest.config.js
Comment (aecorredor) [inline jest.config.js:23]: I'd say we avoid having specific coverage for files and we just up the global coverage based on what we get after these latest changes.

**Review by aecorredor — APPROVED** [review]
My coverage comment is non-blocking, but I think it's worth changing in a follow up. Curious what others think.

File: jest.config.js
Comment (mdlavin) [reply jest.config.js:23]: Absolutely!

File: src/APIBasedAuth.ts
Comment (mdlavin) [inline src/APIBasedAuth.ts:43]: If you are using async/await in the code now, can you also remove the `.then` and `.catch` use?

File: src/APIBasedAuth.ts
Comment (mdlavin) [inline src/APIBasedAuth.ts:196]: Could you simplify this to always return a `Promise`?

File: src/utils/helper.ts
Comment (mdlavin) [inline src/utils/helper.ts:11]: Does this cause a non-`Error` object to be thrown? That's not a very nice pattern.

File: test/APIBasedAuth.test.ts
Comment (mdlavin) [inline test/APIBasedAuth.test.ts:316]: Can you remove the "for coverage" labels here an elsewhere? All tests are added for coverage =)

File: test/APIBasedAuth.test.ts
Comment (mdlavin) [inline test/APIBasedAuth.test.ts:341]: `remoteItem` typo here

File: jest.config.js
Comment (virdesai) [reply jest.config.js:23]: if we want that we're going to have to add coverage to the LOAuth which already exists too and doesn't have 100% coverage. which adds to the scope so wasn't sure that we wanted add that globally

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:196]: @mdlavin if we do that, the default window.localStorage does not fit into the storage type as the `get/set/remove` within that aren't promises

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:43]: we could, we just have to add a try-catch around it too. think that's a worthwhile pattern to follow for any specific reason?

File: src/APIBasedAuth.ts
Comment (mdlavin) [reply src/APIBasedAuth.ts:196]: That's a bummer. I won't be the one to maintain it, but I'd be tempted to wrap `window.localStorage` to have a Promise API just for sake of consistency.

File: src/APIBasedAuth.ts
Comment (mdlavin) [reply src/APIBasedAuth.ts:43]: I think that `try/catch` is easier to understand than `.then`/`.catch`. I think it would be a simpler method after the switch. If you've tried it and you think the method is more complicated with `try/catch` then I'll defer to you

**Review by jkdowdle — DISMISSED** [review]

File: src/APIBasedAuth.ts
Comment (mdlavin) [inline src/APIBasedAuth.ts:47]: The `storageKeys` seems like would be reasonble to provide defaults.

File: src/APIBasedAuth.ts
Comment (mdlavin) [inline src/APIBasedAuth.ts:231]: Should this be made required now?

File: src/utils/helper.ts
Comment (mdlavin) [inline src/utils/helper.ts:11]: Is `error.response?.data` an Error?

File: src/APIBasedAuth.ts
Comment (mdlavin) [inline src/APIBasedAuth.ts:225]: Nice API improvements

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:231]: yeah auth works without storage being set but there's no automatically handled token/session data to the user of app-tools if so. the user still has the ability to pull the values returned from using the functions and store them themselves

File: src/utils/helper.ts
Comment (virdesai) [reply src/utils/helper.ts:11]: error would be an AxiosError, not quite sure what error.response?.data is, just pulling this from https://github.com/PreventiaGroupLLC/ClientUI/blob/master/src/util/formatAxiosError.ts

Comment (smolster) [reply src/utils/helper.ts:11]: We should definitely only throw `Error` objects -- `error.response?.data` won't be that. :( I think the code you linked happens to be a bug.

This is unfair for late-stage feedback: could/should this be re-organized into an `axios` interceptor, so that we don't have to remember to call `formatAxiosError` everywhere?

File: src/APIBasedAuth.ts
Comment (virdesai) [reply src/APIBasedAuth.ts:47]: the default keys are inline so the user has the ability to currently use our default naming for any keys and use their own custom naming for others at their discretion. there are the defaults currently
```
{
  accessToken: 'lo-app-tools-api-auth.accessToken',
  idToken: 'lo-app-tools-api-auth.idToken',
  refreshToken: 'lo-app-tools-api-auth.refreshToken',
  session: 'lo-app-tools-api-auth.session',
  username: 'lo-app-tools-api-auth.username',
}
```
if the users wanted to just have custom values for session/username they could without modifying the default for the other 3, but if we defaulted storageKeys then they would have to provide all 5 keys in the class initialization

File: src/utils/helper.ts
Comment (virdesai) [reply src/utils/helper.ts:11]: haha I originally had it as an axios interceptor and I can put it back as one

File: jest.config.js
Comment (aecorredor) [inline jest.config.js:17]: After these changes, is it safe to remove the specific coverage for `APIBasedAuth.ts`?

File: src/utils/helper.ts
Comment (aecorredor) [reply src/utils/helper.ts:11]: Just to makes sure, we'll we add this to an axios interceptor in this PR or in a follow-up?

**Review by aecorredor — APPROVED** [review]
Not sure which feedback needs to be addressed here or in a follow-up. But I'm good with merging this one and then following up with another one, since the discussion became pretty long and seems harder to follow at this point.

**Review by jkdowdle — APPROVED** [review]

File: src/APIBasedAuth.ts
Comment (smolster) [inline src/APIBasedAuth.ts:11]: Wow. These docs are killer!!

File: src/APIBasedAuth.ts
Comment (smolster) [reply src/APIBasedAuth.ts:47]: I think you're saying: defaults will be used for any unspecified keys. If that's right, I think we're in good shape!

**Review by smolster — APPROVED** [review]
Looks good. Fine to do the axios interceptor update in a subsequent PR.

**Comment by virdesai** [comment]
> I think you're saying: defaults will be used for any unspecified keys. If that's right, I think we're in good shape!

@smolster for some reason I can't quote reply [your message](https://github.com/lifeomic/app-tools/pull/62#discussion_r892526749) but yeah, defaults are used for any unspecified keys and the user of app-tools has the ability to customize { 0, 1, ..., n } keys if they want to

**Pull request merged**
